src/encoding.rs

```diff
@@ -2,69 +2,62 @@
 //! All methods err using VMError::FormatError for convenience.
 
 use byteorder::{ByteOrder, LittleEndian};
+use core::ops::Range;
 use curve25519_dalek::ristretto::CompressedRistretto;
 use curve25519_dalek::scalar::Scalar;
 
 use crate::errors::VMError;
 
-/// Reads a byte and returns the subsequent slice
-pub fn read_u8<'a>(slice: &'a [u8]) -> Result<(u8, &'a [u8]), VMError> {
-    if slice.len() < 1 {
-        return Err(VMError::FormatError);
-    }
-    Ok((slice[0], &slice[1..]))
+pub struct SliceView<'a,T> {
+    whole: &'a [T],
+    range: Range<usize>
 }
 
-/// Reads the LE32-encoded integer and returns the subsequent slice
-pub fn read_u32<'a>(slice: &'a [u8]) -> Result<(u32, &'a [u8]), VMError> {
-    if slice.len() < 4 {
-        return Err(VMError::FormatError);
+impl<'a, T> SliceView<'a,T> {
+    // get returns the slice from the given range, advancing the
+    // internal range
+    fn get(mut self, offset: usize) -> Result<&'a [T], VMError> {
+        let result = self.whole.get(self.range.start..self.range.start+offset).ok_or(VMError::FormatError);
+        self.range.start = self.range.start+offset;
+        result
     }
-    let x = LittleEndian::read_u32(slice);
-    Ok((x, &slice[4..]))
 }
 
-/// Reads the LE32-encoded integer as `usize` and returns the subsequent slice
-pub fn read_usize<'a>(slice: &'a [u8]) -> Result<(usize, &'a [u8]), VMError> {
-    let (n, rest) = read_u32(slice)?;
-    Ok((n as usize, rest))
+pub fn read_next_u8<'a>(slice: &mut SliceView<'a, u8>) -> Result<u8, VMError> {
```

Comment:
We should keep chainable semantics of prev API by returning a reduced sliceview along with parsed object. And since it is Copy, we can move the sliceview in instead of &mut argument.

Reply:
ah, got it! so rather than mutating the `SliceView`, we'll just return a new one w/ the reduced range?

Reply:
Sorry for being unclear. My suggestion was to mutate the sliceview, but instead of taking &mut, just move it in, mutate, then return it. This is what i was doing with normal slices. But your api leads to cleaner code, so lets keep it!